File: python/mirage/core/gmail/stat.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mimetypes import guess_type as _guess_mime

from mirage.accessor.gmail import GmailAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.gmail.labels import list_labels
from mirage.core.gmail.readdir import readdir as _readdir
from mirage.types import FileStat, FileType, PathSpec
from mirage.utils.errors import enoent
from mirage.utils.filetype import filetype_from_mimetype


def _guess_filetype(filename: str) -> FileType:
    mime, _ = _guess_mime(filename)
    return filetype_from_mimetype(mime or "")


async def stat(
    accessor: GmailAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> FileStat:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original

    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    if not key:
        return FileStat(name="/", type=FileType.DIRECTORY)
    if index is None:
        raise enoent(virtual)
    virtual_key = prefix + "/" + key if prefix else "/" + key
    result = await index.get(virtual_key)
    if result.entry is None and "/" not in key:
        labels = await list_labels(accessor.token_manager)
        names = {
            lb["id"] if lb.get("type") == "system" else lb.get(
                "name", lb["id"])
            for lb in labels
        }
        if key in names:
            return FileStat(name=key, type=FileType.DIRECTORY)
        raise enoent(virtual)
    if result.entry is None:
        parent_virtual = virtual_key.rsplit("/", 1)[0] or "/"
        try:
            await _readdir(
                accessor,
                PathSpec(original=parent_virtual,
                         directory=parent_virtual,
                         prefix=prefix),
                index=index,
            )
        # best-effort cache populate; canonical ENOENT raised below
        except Exception:
            pass
        result = await index.get(virtual_key)
        if result.entry is None:
            raise enoent(virtual)
    rt = result.entry.resource_type
    if rt == "gmail/label":
        return FileStat(
            name=result.entry.vfs_name,
            type=FileType.DIRECTORY,
            extra={"label_id": result.entry.id},
        )
    if rt == "gmail/date":
        return FileStat(
            name=result.entry.vfs_name,
            type=FileType.DIRECTORY,
        )
    if rt == "gmail/message":
        return FileStat(
            name=result.entry.vfs_name,
            type=FileType.JSON,
            size=result.entry.size,
            extra={"message_id": result.entry.id},
        )
    if rt == "gmail/attachment_dir":
        return FileStat(
            name=result.entry.vfs_name,
            type=FileType.DIRECTORY,
            extra={"message_id": result.entry.id},
        )
    if rt == "gmail/attachment":
        ft = _guess_filetype(result.entry.vfs_name)
        return FileStat(
            name=result.entry.vfs_name,
            type=ft,
            size=result.entry.size,
            extra={"attachment_id": result.entry.id},
        )
    return FileStat(
        name=result.entry.vfs_name,
        type=FileType.JSON,
        extra={"message_id": result.entry.id},
    )

File: typescript/packages/core/src/core/gmail/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GmailAccessor } from '../../accessor/gmail.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { FileStat, FileType, PathSpec } from '../../types.ts'
import { readdir as coreReaddir } from './readdir.ts'
import { listLabels } from './labels.ts'
import { stripSlash } from '../../utils/slash.ts'
import { enoent } from '../../utils/errors.ts'

function guessType(name: string): FileType {
  const lower = name.toLowerCase()
  if (lower.endsWith('.json') || lower.endsWith('.gmail.json')) return FileType.JSON
  if (lower.endsWith('.csv')) return FileType.CSV
  if (lower.endsWith('.png')) return FileType.IMAGE_PNG
  if (lower.endsWith('.jpg') || lower.endsWith('.jpeg')) return FileType.IMAGE_JPEG
  if (lower.endsWith('.gif')) return FileType.IMAGE_GIF
  if (lower.endsWith('.zip')) return FileType.ZIP
  if (lower.endsWith('.gz') || lower.endsWith('.gzip')) return FileType.GZIP
  if (lower.endsWith('.pdf')) return FileType.PDF
  if (lower.endsWith('.txt') || lower.endsWith('.md') || lower.endsWith('.log')) {
    return FileType.TEXT
  }
  return FileType.BINARY
}

export async function stat(
  accessor: GmailAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<FileStat> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (key === '') return new FileStat({ name: '/', type: FileType.DIRECTORY })

  if (index === undefined) throw enoent(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`
  let result = await index.get(virtualKey)
  if (result.entry === undefined || result.entry === null) {
    if (!key.includes('/')) {
      const labels = await listLabels(accessor.tokenManager)
      const names = new Set(labels.map((lb) => (lb.type === 'system' ? lb.id : (lb.name ?? lb.id))))
      if (names.has(key)) return new FileStat({ name: key, type: FileType.DIRECTORY })
      throw enoent(path.original)
    }
    const parentVirtual = virtualKey.slice(0, virtualKey.lastIndexOf('/')) || '/'
    try {
      await coreReaddir(
        accessor,
        new PathSpec({
          original: parentVirtual,
          directory: parentVirtual,
          resolved: false,
          prefix,
        }),
        index,
      )
    } catch {
      // parent listing failed — fall through
    }
    result = await index.get(virtualKey)
    if (result.entry === undefined || result.entry === null) {
      throw enoent(path.original)
    }
  }
  const rt = result.entry.resourceType
  const vfsName = result.entry.vfsName !== '' ? result.entry.vfsName : result.entry.name
  if (rt === 'gmail/label') {
    return new FileStat({
      name: vfsName,
      type: FileType.DIRECTORY,
      extra: { label_id: result.entry.id },
    })
  }
  if (rt === 'gmail/date') {
    return new FileStat({ name: vfsName, type: FileType.DIRECTORY })
  }
  if (rt === 'gmail/message') {
    return new FileStat({
      name: vfsName,
      type: FileType.JSON,
      size: result.entry.size,
      extra: { message_id: result.entry.id },
    })
  }
  if (rt === 'gmail/attachment_dir') {
    return new FileStat({
      name: vfsName,
      type: FileType.DIRECTORY,
      extra: { message_id: result.entry.id },
    })
  }
  if (rt === 'gmail/attachment') {
    return new FileStat({
      name: vfsName,
      type: guessType(vfsName),
      size: result.entry.size,
      extra: { attachment_id: result.entry.id },
    })
  }
  return new FileStat({
    name: vfsName,
    type: FileType.JSON,
    extra: { message_id: result.entry.id },
  })
}
